silence/silence_test.go

```diff
@@ -84,29 +84,130 @@ func TestOptionsValidate(t *testing.T) {
 	}
 }
 
-func TestSilencesGC(t *testing.T) {
-	s, err := New(Options{})
-	require.NoError(t, err)
-
-	s.clock = clock.NewMock()
-	now := s.nowUTC()
-
-	newSilence := func(exp time.Time) *pb.MeshSilence {
-		return &pb.MeshSilence{ExpiresAt: exp}
+func TestSilenceGCOverTime(t *testing.T) {
+	type silenceEntry struct {
+		s                    *pb.Silence
+		expectPresentAfterGc bool
 	}
-	s.st = state{
-		"1": newSilence(now),
-		"2": newSilence(now.Add(-time.Second)),
-		"3": newSilence(now.Add(time.Second)),
+	c := clock.NewMock()
+	now := c.Now().UTC()
+
+	newSilence := func(id string, exp time.Time) *pb.Silence {
+		return &pb.Silence{
+			Id:       id,
+			StartsAt: now.Add(-time.Second),
+			EndsAt:   exp,
+			Matchers: []*pb.Matcher{
+				{Name: "foo", Type: pb.Matcher_REGEXP, Pattern: "bar"},
+			}}
 	}
-	want := state{
-		"3": newSilence(now.Add(time.Second)),
+
+	// The GC will run with it's clock equal to 1 second after now
+	cases := []struct {
+		name            string
+		initialState    []silenceEntry
+		updates         []silenceEntry
+		expectedGCCount int
+	}{
+		{
+			name: "gc does not clean active silences",
+			initialState: []silenceEntry{
+				{s: newSilence("1", now), expectPresentAfterGc: false},
+				{s: newSilence("2", now.Add(-time.Second)), expectPresentAfterGc: false},
+				{s: newSilence("3", now.Add(time.Second)), expectPresentAfterGc: true},
+			},
+		},
+		{
+			name: "silences added with Set are handled correctly",
```

Comment:
> handled correctly

What does this mean?

Reply:
Ah this case name isn't great - in this case "handled correctly" just means "all the invariants we're testing remain satisfied"

Reply:
> "all the invariants we're testing remain satisfied"

Yeah but what are the invariants being tested? I'm sort of coming at this from the perspective of someone who hasn't reviewed this PR but needs to understand the tests. It's **so hard** to understand **what** is being tested here from looking at the test case:

```go
{
	name: "silences added with Set are handled correctly",
	initialState: []silenceEntry{
		{s: newSilence("1", now), expectPresentAfterGc: false},
	},
	updates: []silenceEntry{
		{s: newSilence("", now.Add(time.Second)), expectPresentAfterGc: true},
		{s: newSilence("", now.Add(-time.Second)), expectPresentAfterGc: false},
	},
},
```